subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/IvyArtifactNotationParserFactory.java

```diff
@@ -104,7 +105,7 @@ private FileNotationConverter(FileResolver fileResolver) {
         public void convert(Object notation, NotationConvertResult<? super IvyArtifact> result) throws TypeConversionException {
             File file = fileResolverNotationParser.parseNotation(notation);
             IvyArtifact ivyArtifact = instantiator.newInstance(FileBasedIvyArtifact.class, file, publicationIdentity);
-            if (notation instanceof TaskDependencyContainer) {
+            if (dependsOnTask(notation)) {
                 ivyArtifact.builtBy(notation);
```

Comment:
Adding a `Provider` as a task dependency to a `DefaultTaskDependency` (via `builtBy()`) has a special meaning, where if the provider does not represent a task output, then the value is assumed to be something that can be resolved to a task dependency. This isn't what we want here.

Ideally, we would parse the notation to a `Provider<File>` to take care of all this and get rid of the 'builtBy' entirely. For a quick fix, I think we can do something like (this is off the top of my head):

```
if (notation instanceof TaskDependencyContainer) {
    ivyArtifact.buildBy(new TaskDependencyContainer() {
        void visitDependencies(TaskDependencyResolveContext context) {
            context.add(notation);
        }
    })
}
```

Here, we're wrapping the notation in a `TaskDependencyContainer` to hide the provider from the special case handling in `DefaultTaskDependency`.

Reply:
Thanks. That's working. I pushed the change.

I think this is ok for now and we can evolve it later. We now have the test to make sure we do not break the use case again. This is also a "discouraged" part of the API, as it is preferred to add artifacts through configurations/variants and components.